File: tracks/python/list_ops.py
def append(list1, list2):
    for item in list2:
        list1.append(item)
    return list1


def concat(lists):
    result = []
    for list in lists:
        for item in list:
            if item not in result:
                result.append(item)
    return result


def filter(function, list):
    result = []
    for item in list:
        if function(item):
            result.append(item)
    return result


def length(list):
    count = 0
    for item in list:
        count += 1
    return count


def map(function, list):
    result = []
    for item in list:
        result.append(function(item))
    return result


def foldl(function, list, initial):
    acc = initial
    for element in list:
        acc = function(acc, element)
    return acc


def foldr(function, list, initial):
    return foldl(function, list[::-1], initial)


def reverse(list):
    result = []
    for item in list[::-1]:
        result.append(item)
    return result

File: tracks/swift/ListOps.swift
//Solution goes in Sources
func append(_ lhs: [Int], _ rhs: [Int]) -> [Int] {
    var result = lhs
    for val in rhs {
        result.append(val)
    }
    return result
}

func concat(_ lists: [Int]...) -> [Int] {
    var result: [Int] = []
    for list in lists {
        result += list
    }
    return result
}

func filter(_ list: [Int], function f: (Int) -> Bool) -> [Int] {
    var result: [Int] = []
    for val in list {
        if f(val) {
            result.append(val)
        }
    }
    return result
}

func length(_ list: [Int]) -> Int {
    var result = 0
    for _ in list {
        result += 1
    }
    return result
}

func map(_ list: [Int], function f: (Int) -> Int) -> [Int] {
    var result: [Int] = []
    for val in list {
        result.append(f(val))
    }
    return result
}

func foldLeft<T>(_ list: ArraySlice<T>, accumulated: T, combine f: (T, T) -> T) -> T {
    if let first = list.first {
        return foldLeft(list.dropFirst(1), accumulated: f(accumulated, first), combine: f)
    } else {
        return accumulated
    }
}

func foldRight<T>(_ list: ArraySlice<T>, accumulated: T, combine f: (T, T) -> T) -> T {
    if let first = list.first {
        return f(first, foldRight(list.dropFirst(1), accumulated: accumulated, combine: f))
    } else {
        return accumulated
    }
}

func reverse(_ list: [Int]) -> [Int] {
    var result = list
    let n = list.count
    let upTo = n / 2
    for i in 0..<upTo {
        let tmp = result[i]
        result[i] = result[n - i - 1]
        result[n - i - 1] = tmp
    }
    return result
}